src/io/dataset.cpp

```diff
@@ -521,13 +520,64 @@ MultiValBin* Dataset::GetMultiBinFromSparseFeatures() const {
              sum_sparse_rate);
   std::unique_ptr<MultiValBin> ret;
   ret.reset(MultiValBin::CreateMultiValBin(num_data_, offsets.back(),
-                                           num_feature, sum_sparse_rate));
+                                           num_feature, sum_sparse_rate, offsets));
   PushDataToMultiValBin(num_data_, most_freq_bins, offsets, &iters, ret.get());
   ret->FinishLoad();
   return ret.release();
 }
 
-MultiValBin* Dataset::GetMultiBinFromAllFeatures() const {
+MultiValBin* Dataset::GetMultiBinFromDenseFeatures(const std::vector<uint32_t>& offsets) const {
+  Common::FunctionTimer fun_time("Dataset::GetMultiBinFromDenseFeatures",
+                                 global_timer);
+  int num_threads = OMP_NUM_THREADS();
+  double sum_dense_ratio = 0.0f;
+
+  std::unique_ptr<MultiValBin> ret;
+  std::vector<std::vector<std::unique_ptr<BinIterator>>> iters(num_threads);
+  std::vector<uint32_t> most_freq_bins;
+  std::vector<uint32_t> local_offsets;
+  int ncol = 0;
+  for (int gid = 0; gid < num_groups_; ++gid) {
+    if (feature_groups_[gid]->is_multi_val_) {
+      continue;
+    }
+    ncol += 1;
+    for (int fid = 0; fid < feature_groups_[gid]->num_feature_; ++fid) {
+      const auto& bin_mapper = feature_groups_[gid]->bin_mappers_[fid];
+      sum_dense_ratio += 1.0f - bin_mapper->sparse_rate();
+    }
+  }
+  sum_dense_ratio /= ncol;
+  const int offset = (1.0f - sum_dense_ratio) >=
+    MultiValBin::multi_val_bin_sparse_threshold ? 1 : 0;
+  int num_total_bin = offset;
+  local_offsets.push_back(num_total_bin);
+  for (int gid = 0; gid < num_groups_; ++gid) {
+    if (!feature_groups_[gid]->is_multi_val_) {
+      most_freq_bins.push_back(0);
+      num_total_bin += feature_groups_[gid]->bin_offsets_.back() - offset;
+      for (int tid = 0; tid < num_threads; ++tid) {
+        iters[tid].emplace_back(feature_groups_[gid]->FeatureGroupIterator());
+      }
+      local_offsets.push_back(num_total_bin);
+    }
+  }
+  CHECK(local_offsets.size() == offsets.size());
+  for (size_t i = 0; i < local_offsets.size(); ++i) {
+    CHECK(local_offsets[i] == offsets[i]);
```

Comment:
what is `local_offsets` used for? only for checking?

Reply:
the passed-in offset is already the offset of dense features? why need to re-campute local_offset ?

Reply:
This is just for checking.